applications/sdp/mspi/src/main.c

```diff
@@ -516,10 +517,106 @@ __attribute__((interrupt)) void hrt_handler_write(void)
 	hrt_write((hrt_xfer_t *)&xfer_params);
 }
 
+/**
+ * @brief Trap handler for SDP application.
+ *
+ * @details
+ * This function is called on unhandled CPU exceptions. It's a good place to
+ * handle critical errors and notify the core that the SDP application has
+ * crashed.
+ *
+ * @param mcause  - cause of the exception (from mcause register)
+ * @param mepc    - address of the instruction that caused the exception (from mepc register)
+ * @param mtval   - additional value (e.g. bad address)
+ * @param context - pointer to the saved context (only some registers are saved - ra, t0, t1, t2)
+ */
+void trap_handler(uint32_t mcause, uint32_t mepc, uint32_t mtval, void *context)
+{
+	volatile const uint8_t fault_opcode = NRFE_MSPI_SDP_APP_HARD_FAULT;
+	volatile uint32_t cause = mcause;
+	volatile uint32_t pc = mepc;
+	volatile uint32_t bad_addr = mtval;
+	volatile uint32_t *ctx = (volatile uint32_t *)context;
+
+	/* It can be distinguish whether the exception was caused by an interrupt or an error:
+	 * On RV32, bit 31 of the mcause register indicates whether the event is an interrupt.
+	 */
+	if (cause & 0x80000000) {
+		/* Interrupt – can be handled or ignored */
+	} else {
+		/* Exception – critical error */
+	}
+
+	cpuflpr_error_ctx_ptr[0] = cause;
+	cpuflpr_error_ctx_ptr[1] = pc;
+	cpuflpr_error_ctx_ptr[2] = bad_addr;
+	cpuflpr_error_ctx_ptr[3] = (uint32_t)ctx;
+
+	ipc_service_send(&ep, (const uint8_t *)&fault_opcode, sizeof(fault_opcode));
+
+	while (1) {
+	}
+}
+
+/* The trap_entry function is the entry point for exception handling.
+ * The naked attribute prevents the compiler from generating an automatic prologue/epilogue.
+ */
+__attribute__((naked)) void trap_entry(void)
+{
+	__asm__ volatile(
+		/* Reserve space on the stack:
+		 * 16 bytes for 4 registers context (ra, t0, t1, t2).
+		 */
+		"addi sp, sp, -16\n"
+		"sw   ra, 12(sp)\n"
+		"sw   t0, 8(sp)\n"
+		"sw   t1, 4(sp)\n"
+		"sw   t2, 0(sp)\n"
+
+		/* Read CSR: mcause, mepc, mtval */
+		"csrr t0, mcause\n" /* t0 = mcause */
+		"csrr t1, mepc\n"   /* t1 = mepc   */
+		"csrr t2, mtval\n"  /* t2 = mtval  */
+
+		/* Prepare arguments for trap_handler function:
+		 * a0 = mcause (t0), a1 = mepc (t1), a2 = mtval (t2), a3 = sp (pointer on context).
+		 */
+		"mv   a0, t0\n"
+		"mv   a1, t1\n"
+		"mv   a2, t2\n"
+		"mv   a3, sp\n"
+
+		"call trap_handler\n"
+
+		/* Restore registers values */
+		"lw   ra, 12(sp)\n"
+		"lw   t0, 8(sp)\n"
+		"lw   t1, 4(sp)\n"
+		"lw   t2, 0(sp)\n"
+		"addi sp, sp, 16\n"
+
+		"mret\n");
+}
+
+void init_trap_handler(void)
+{
+	/* Write the address of the trap_entry function into the mtvec CSR.
+	 * Note: On RV32, the address must be aligned to 4 bytes.
+	 */
+	uintptr_t trap_entry_addr = (uintptr_t)&trap_entry;
+
+	__asm__ volatile("csrw mtvec, %0\n"
+			 : /* no outs */
+			 : "r"(trap_entry_addr));
```

Comment:
I guess it should be noted somewhere, that SDP overwrites the default trap handler. Maybe it should be a KConfig option enabled by default?